lib/ExecutionEngine/LEVELZERORUNTIME/LevelZeroRuntimeWrappers.cpp

```diff
@@ -0,0 +1,315 @@
+  }
+}
+
+inline void checkResult(ze_result_t res, const char *func) {
+  if (res != ZE_RESULT_SUCCESS)
+    throw std::runtime_error(std::string(func) +
+                             " failed: " + std::to_string(res));
+}
+
+#define CHECK_ZE_RESULT(expr) checkResult((expr), #expr)
+
+} // namespace
+
+struct ParamDesc {
+  void *data;
+  size_t size;
+
+  bool operator==(const ParamDesc &rhs) const {
+    return data == rhs.data && size == rhs.size;
+  }
+
+  bool operator!=(const ParamDesc &rhs) const { return !(*this == rhs); }
+};
+
+template <typename T> size_t countUntil(T *ptr, T &&elem) {
+  assert(ptr);
+  auto curr = ptr;
+  while (*curr != elem) {
+    ++curr;
+  }
+  return static_cast<size_t>(curr - ptr);
+}
+
+static std::pair<ze_driver_handle_t, ze_device_handle_t>
+getDriverAndDevice(ze_device_type_t deviceType = ZE_DEVICE_TYPE_GPU) {
+
+  CHECK_ZE_RESULT(zeInit(ZE_INIT_FLAG_GPU_ONLY));
+  uint32_t driverCount = 0;
+  CHECK_ZE_RESULT(zeDriverGet(&driverCount, nullptr));
+
+  std::vector<ze_driver_handle_t> allDrivers{driverCount};
+  CHECK_ZE_RESULT(zeDriverGet(&driverCount, allDrivers.data()));
+
+  // Find a driver instance with a GPU device
+  std::vector<ze_device_handle_t> devices;
+  for (uint32_t i = 0; i < driverCount; ++i) {
+    uint32_t deviceCount = 0;
+    CHECK_ZE_RESULT(zeDeviceGet(allDrivers[i], &deviceCount, nullptr));
+    if (deviceCount == 0)
+      continue;
+    devices.resize(deviceCount);
+    CHECK_ZE_RESULT(zeDeviceGet(allDrivers[i], &deviceCount, devices.data()));
+    for (uint32_t d = 0; d < deviceCount; ++d) {
+      ze_device_properties_t device_properties;
+      CHECK_ZE_RESULT(zeDeviceGetProperties(devices[d], &device_properties));
+      if (deviceType == device_properties.type) {
+        auto driver = allDrivers[i];
+        auto device = devices[d];
+        return {driver, device};
+      }
+    }
+  }
+  throw std::runtime_error("getDevice failed");
+}
+
+struct GPUL0QUEUE {
+
+  ze_driver_handle_t zeDriver_ = nullptr;
+  ze_device_handle_t zeDevice_ = nullptr;
+  ze_context_handle_t zeContext_ = nullptr;
+  ze_command_list_handle_t zeCommandList_ = nullptr;
+
+  GPUL0QUEUE() {
+    auto driverAndDevice = getDriverAndDevice();
+    zeDriver_ = driverAndDevice.first;
+    zeDevice_ = driverAndDevice.second;
+
+    ze_context_desc_t contextDesc = {ZE_STRUCTURE_TYPE_CONTEXT_DESC, nullptr,
+                                     0};
+    CHECK_ZE_RESULT(zeContextCreate(zeDriver_, &contextDesc, &zeContext_));
+
+    ze_command_queue_desc_t desc = {};
+    desc.mode = ZE_COMMAND_QUEUE_MODE_ASYNCHRONOUS;
+    CHECK_ZE_RESULT(zeCommandListCreateImmediate(zeContext_, zeDevice_, &desc,
+                                                 &zeCommandList_));
+  }
+  GPUL0QUEUE(ze_device_type_t *deviceType, ze_context_handle_t context) {
+    auto driverAndDevice = getDriverAndDevice(*deviceType);
+    zeDriver_ = driverAndDevice.first;
+    zeDevice_ = driverAndDevice.second;
+
+    zeContext_ = context;
```

Comment:
this context handle needs to be checked for null?

Reply:
it gets checked before this constructor is called from gpuCreateStream